R/calcGAINS2025forREMIND.R

```diff
@@ -0,0 +1,177 @@
+
+    # Drop sectors not in mapping
+    dropsectors <- getItems(innew, "sector")[!(getItems(innew, "sector") %in% secmap$gains)]
+    # Also drop sectors that are not mapped with sufficient detail. Those that can be
+    # used will have a dot "." splitting the specific technologies
+    dropsectors <- c(secmap$gains[!grepl("\\.", secmap$remind)], dropsectors)
+    innew <- innew[, , dropsectors, invert = T]
+    wgtnew <- wgtnew[, , dropsectors, invert = T]
+
+    # Aggregate to REMIND sectors
+    filtsecmap <- secmap[!(secmap$gains %in% dropsectors), ]
+    outnew <- toolAggregate(innew, filtsecmap, weight = NULL, from = "gains", to = "remind", dim = "sector")
+    wgtnew <- toolAggregate(wgtnew, filtsecmap, weight = NULL, from = "gains", to = "remind", dim = "sector")
+
+    # Use abind to split the dimensions of specific technologies into subsectors
+    splitTechs <- function(mag) {
+      hasssp <- "ssp" %in% getSets(mag)
+      mag <- as.magpie(abind::abind(clean_magpie(mag)), spatial = 1, temporal = 2)
+      splitsectors <- c("region", "year", "sector1", "sector2", "sector3", "sector4", "emi", "scenario")
+      if (hasssp) splitsectors <- c(splitsectors, "ssp")
+      getSets(mag) <- splitsectors
+      return(mag)
+    }
+    outnew <- splitTechs(outnew)
+    wgtnew <- splitTechs(wgtnew)
+
+    # Drop sectors not used in REMIND anymore
+    outnew <- outnew[, , c("pcc", "pco"), invert = T]
+    wgtnew <- wgtnew[, , c("pcc", "pco"), invert = T]
+    # ==============================================================================================================
+    # GAINSlegacy
+    # ==============================================================================================================
+    # GAINSlegacy does not actually follow its stated dimensions, so they are meaningless from this point on
+    linold <- calcOutput(
+      "EmissionFactors",
+      subtype = "emission_factors", warnNA = FALSE, aggregate = F, supplementary = T
+    )
+    inold <- linold$x
+    wgtold <- linold$w
+
+    # Split dimensions
+    outold <- splitTechs(inold)
+    wgtold <- splitTechs(wgtold)
+
+    # Pad with SSP dimension
+    nextdim <- as.numeric(paste0("3.", length(getSets(outold)) - 1))
+    outold <- add_dimension(outold, nextdim, add = "ssp", nm = "GAINSlegacy")
+    wgtold <- add_dimension(wgtold, nextdim, add = "ssp", nm = "GAINSlegacy")
+
+    # GAINSlegacy weights were for a fixed year, expand the time dimension
+    # so that it can be merged with the new data
+    wgtold <- mbind(lapply(getYears(wgtnew), \(x) setYears(wgtold, x)))
+
+    # ==============================================================================================================
+    # Combining GAINSlegacy and GAINS2025
+    # ==============================================================================================================
+    out <- mbind(outold, outnew)
+    wgt <- mbind(wgtold, wgtnew)
```

Comment:
The chosen weight produces the warning `Weight sum is 0, so cannot normalize and will return 0 for some aggregation targets. This changes the total sum of the magpie object! If this is really intended set zeroWeight = "allow", or "setNA" to return NA.`

Basically, what happens is that when summing up the weights for some regions, the weigtht sum is zero and therefore, values that were not zero can get lost in the process of regional aggregation. Is this ok for you?